DRepDistr: Iterate over the DRep delegations in UMap

aniketd opened the pull request:
# Description

Resolves #4233 

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

aniketd committed:
DRepDistr: iterate over UMap in chunks.

The DRepDistr was being calculated by iterating over the StakeDistr,
which meant that delegations that have only rewards and no UTxOs would
be left out. This commit changes the calculation to iterate over the
UMap in chunks instead, adding the corresponding rewards and StakeDistr
in the calculation and producing a DRepDistr for all DReps that are
either registered or one of the eternal ones, irrespective of whether
they have an associated StakeDistr.

Also add `UMap.umElemDRepDelegatedReward` to get only those rewards that
have an active DRep delegation.

aniketd commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs:
```diff
@@ -136,7 +136,7 @@ module Cardano.Ledger.Conway.Governance (
   forceDRepPulsingState,
   finishDRepPulser,
-  computeDrepDistr,
+  computeDRepDistr,
```
Needs changelog entry

aniketd commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -292,6 +293,30 @@ registerDRep = do
   dreps `shouldSatisfy` Map.member (KeyHashObj khDRep)
   pure khDRep
 
+setupDRepWithoutUTxO ::
```
Needs changelog entry

aniketd commented on libs/cardano-ledger-core/src/Cardano/Ledger/UMap.hs:
```diff
@@ -262,6 +263,20 @@ umElemAsTuple = \case
   TFFFF r p s v -> (SJust r, p, SJust s, SJust v)
 {-# INLINE umElemAsTuple #-}
 
+-- | Extract a DRep delegated reward-deposit pair if it is present.
```
Needs changelog entry and version bump

aniketd reviewed (commented):
Self-aware TODOs

teodanciu commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -292,6 +293,30 @@ registerDRep = do
+              (DelegVote (DRepCredential $ KeyHashObj drepKH))
+              zero
+          ]
+  pure (drepKH, delegatorKH)
```
~Is it the case that having an extra txOut changes the rewards? I think when not explicitly set, the root will be the txOut, so it's not technically "withoutUtxo". 
The name of the function is a little confusing in my opinion, and also kinda specific.
But maybe because I'm missing the semantics of it - why do we actually need it like this, what's the consequences?  If possible, I would either try to find a better name (to reflect the intention) or if not maybe move it to the spec itself?~

teodanciu replied:
Ok, after looking more at the PR, I understand  now the intention. 
Still I think the name is a little confusing, but  I have no better idea. Maybe you could add a comment explaining the difference between this and `setupSingleDRep`   (I suppose that being, that the Drep won't have no moneyz)

aniketd replied:
Sure! And no worries! :sweat_smile:  Thanks!

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance/DRepPulser.hs:
```diff
@@ -160,60 +159,39 @@ instance EraPParams era => ToCBOR (PulsingSnapshot era) where
+              DRepAlwaysNoConfidence -> distrWithStakeAndReward
+              DRepCredential drepCred
+                | Map.member drepCred regDReps -> distrWithStakeAndReward
+                | otherwise -> dRepDistr
```
I wonder if this shouldn't be `distr`  instead of `dRepDistr`, since it's the accumulator value within the `accum` function, pertaining to it's sibling `drep` argument.

aniketd replied:
Awesome catch! Thank you! :raised_hands:

aniketd commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance/DRepPulser.hs:
```diff
@@ -160,60 +159,51 @@ instance EraPParams era => ToCBOR (PulsingSnapshot era) where
+--    + (m * log((n/m) + 1)) where m <= n -- union stakedistr with active-rewards
+--    + (size of UMap) -- filter from UMap all dreps
+--    + (size of dreps * log (size of stakesAndRewards)) -- lookup stakes for each drep
+-- )
```
@teodanciu @Soupstraw can I get your review on this calculation?

aniketd commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance/DRepPulser.hs:
```diff
@@ -160,60 +159,51 @@ instance EraPParams era => ToCBOR (PulsingSnapshot era) where
---   3) stakeDistr :: VMap VB VP (Credential 'Staking c) (CompactForm Coin). The
---   aggregated stake distr extracted from the first component of the IncrementalStake
---   i.e. (IStake credmap _) where credmap is converted to a VMap
+-- In contrast to `setupSingleDRep`, this function does not make a UTxO entry
```
This is to be shifted to the `ImpTest`

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance/DRepPulser.hs:
```diff
@@ -160,60 +159,47 @@ instance EraPParams era => ToCBOR (PulsingSnapshot era) where
+          stakeDistr
+          $ fmap rdReward
+          $ Map.mapMaybe umElemRDDRepActive
+          $ umElems um
```
This will totally negate the whole purpose of the pulser.

The reason why we have a pulser, so we don't iterate over the UMap all at once, which is exactly what you are doing here.

lehins reviewed (changes requested):
Unfortunately your solution will not work.

We need to figure out a different one.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/Instances.hs:
```diff
@@ -1213,7 +1213,7 @@ instance IsConwayUniv fn => HasSpec fn (PulsingSnapshot (ConwayEra StandardCrypt
   '[ Int
    , UMap StandardCrypto
-   , Map (Credential 'Staking StandardCrypto) (CompactForm Coin)
+   , Int -- Map (Credential 'Staking StandardCrypto) (CompactForm Coin)
```
```suggestion
   , Int

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs:
```diff
@@ -561,6 +613,57 @@ votingSpec =
+              %~ UM.adjust
+                (\(UM.RDPair r d) -> UM.RDPair (r <> UM.CompactCoin 1_000_000) d)
+                delegatorCStaking2
+                . UM.RewDepUView
```
Ditto

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs:
```diff
@@ -561,6 +613,57 @@ votingSpec =
+              %~ UM.adjust
+                (\(UM.RDPair r d) -> UM.RDPair (r <> UM.CompactCoin 200_000) d)
+                delegatorCStaking1
+                . UM.RewDepUView
```
Ditto
```suggestion
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs:
```diff
@@ -487,6 +487,58 @@ votingSpec =
+              %~ UM.adjust
+                (\(UM.RDPair r d) -> UM.RDPair (r <> UM.CompactCoin 1_000_000) d)
+                (KeyHashObj stakingKH1)
+                . UM.RewDepUView
```
We should not be modifying the state manually in the imp test framework.

What is the way this can happen in production? Easy: submit a proposal and wait for it to expire. This will put the refund into the reward account.
Alternatively register and retire a stake pool will achieve the same: refund placed into the reward account
```suggestion
          -- Add rewards to delegation #1
```

aniketd replied:
Done! Let me know what you think of the way I have done it :slightly_smiling_face:

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs:
```diff
@@ -487,6 +487,58 @@ votingSpec =
+              %~ UM.adjust
+                (\(UM.RDPair r d) -> UM.RDPair (r <> UM.CompactCoin 1_000_000) d)
+                (KeyHashObj stakingKH2)
+                . UM.RewDepUView
```
Ditto
```suggestion
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs:
```diff
@@ -487,6 +487,58 @@ votingSpec =
+              %~ UM.adjust
+                (\(UM.RDPair r d) -> UM.RDPair (r <> UM.CompactCoin 200_000) d)
+                (KeyHashObj stakingKH1)
+                . UM.RewDepUView
```
Ditto
```suggestion
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:
```diff
@@ -292,6 +294,33 @@ registerDRep = do
+          [ mkRegDepositDelegTxCert @era
+              (KeyHashObj delegatorKH)
+              (DelegVote (DRepCredential $ KeyHashObj drepKH))
+              zero
```
We should lookup up the deposit in PParmas similar to `setupPoolWithoutStake`. I'd like to make existing tests a bit more realistic in close future by setting more of the PParams at the beginning of all tests.
```suggestion
              deposit
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/RatifySpec.hs:
```diff
@@ -561,6 +613,57 @@ votingSpec =
+              %~ UM.adjust
+                (\(UM.RDPair r d) -> UM.RDPair (r <> UM.CompactCoin 1_000_000) d)
+                delegatorCStaking1
+                . UM.RewDepUView
```
Ditto
```suggestion
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance/DRepPulser.hs:
```diff
@@ -257,16 +236,16 @@ instance EraPParams era => Eq (DRepPulser era Identity (RatifyState era)) where
-        let !(!steps, !balance') = Map.splitAt dpPulseSize dpBalance
-            drep' = Map.foldlWithKey' (accumDRepDistr dpUMap dpDRepState) dpDRepDistr steps
-         in pure (pulser {dpBalance = balance', dpDRepDistr = drep'})
+        let !chunk = Map.drop dpIndex $ Map.take (dpIndex + dpPulseSize) $ UMap.umElems dpUMap
```
This way it is a bit simpler:
```suggestion
        let !chunk = Map.take dpPulseSize $ Map.drop dpIndex $ UMap.umElems dpUMap
```

lehins commented on eras/conway/impl/CHANGELOG.md:
```diff
@@ -49,6 +51,7 @@
 
 ### `testlib`
 
+* Add `setupDRepWithoutStake` to `Conway.ImpTest`. #4273
```
```suggestion
* Add `setupDRepWithoutStake` and `setupPoolWithoutStake` to `Conway.ImpTest`. #4273
```

lehins reviewed (approved):
Looks great! Tests could use a little improvement, but nothing that would prevent this PR from being merged

aniketd committed:
Address review comments.

Make rewards a result of proposal expiry and pool retirement as it
should be, without directly modified the NewEpochState

lehins commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:
```diff
@@ -1330,6 +1328,38 @@ registerPool = do
+          , ppId = poolKH
+          , ppRewardAccount = RewardAccount networkId stakingC
+          , ppPledge = Coin $ fromIntegral expectedReward
+          , ppCost = zero -- Coin expectedReward
```
Neither pledge nor cost have anything to do with the deposit and the amount of money that will be in the reward account after the pool retires, so these two are unnecessary
```suggestion
  Credential 'Staking (EraCrypto era) ->
  ImpTestM era ()
registerAndRetirePoolToMakeReward stakingC = do
  poolKH <- freshKeyHash
  networkId <- use (to impGlobals . to networkId)
  vrfKH <- freshKeyHashVRF
  Positive pledge <- arbitrary
  Positive cost <- arbitrary
  let poolParams =
        PoolParams
          { ppVrf = vrfKH
          , ppId = poolKH
          , ppRewardAccount = RewardAccount networkId stakingC
          , ppPledge = Coin pledge
          , ppCost = Coin cost
```

aniketd replied:
:bow: Thank you!

lehins reviewed (approved):
Some conflicts resolved and one minor change is still needed, but otherwise it is good to go

aniketd committed:
Merge branch 'master' into aniketd/rewards-in-drep-distr
